### PR TITLE
Added LKQD prebid adapter with associated documentation and tests

### PR DESCRIPTION
**Pull request by mrcrawfo**
## Type of change

- [ ] Bugfix
- [ ] Feature
- [x] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change

- test parameters for validating bids
```
{
  bidder: '<bidder name>',
  params: {
    // ...
  }
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- contact email of the adapter’s maintainer
- [ ] official adapter submission

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:

- A link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/

### CODE REVIEW

**Comment by mkendall07** [comment]
This pull request **introduces 2 alerts** when merging 9bf26824401f0558e9792d305090505fc4082d9e into e6d1fa81f3280633745a6c038274742ef746dc87 - [view on lgtm.com](https://lgtm.com/projects/g/prebid/Prebid.js/rev/pr-f83b88c8d8a06e999c2c97ecf5c70f6a9e13247a)

**new alerts:**

* 2 for Useless conditional

---

*Comment posted by [lgtm.com](https://lgtm.com)*

File: modules/lkqdBidAdapter.js [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+  var bidResponses = [];
+  if (serverResponse && serverResponse.body) {
+    if (serverResponse.error) {
+      utils.logError('Error: ' + serverResponse.error);
+      return bidResponses;
+    } else {
+      try {
+        var bidResponse = {};
+        if (bidRequest && bidRequest.data && bidRequest.data.bidId && bidRequest.data.bidId !== '') {
+          var sspXml = new window.DOMParser().parseFromString(serverResponse.body, 'text/xml');
```
Comment (jaiminpanchal27) [inline modules/lkqdBidAdapter.js]: What happens if this returns error. I think line 175, 179, 180 will throw error. 
https://developer.mozilla.org/en-US/docs/Web/API/DOMParser

Also Can you also confirm your unit tests are passing in Mobile Safari. I am not sure about its support

File: modules/lkqdBidAdapter.js [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        data: sspData
+      });
+    }
+  }
+
+  return bidRequests;
+}
+
+function interpretResponse(serverResponse, bidRequest) {
+  var bidResponses = [];
```
Comment (jaiminpanchal27) [inline modules/lkqdBidAdapter.js]: Use let/const at all places

**Comment by mrcrawfo** [comment]
Thank you for taking the time to review my pull request, I have just pushed
another commit to fix the issues you highlighted with one exception - with
regard to Mobile Safari, I am unclear how to run the unit tests for a
single particular browser. The instructions on
https://github.com/prebid/Prebid.js/blob/master/CONTRIBUTING.md only go as
far as to direct "gulp test", would you be able to clarify what command or
arguments I should be using for the test you are requesting?

Thank you,
Max Crawford
mrcrawfo@gmail.com

**Comment by mrcrawfo** [comment]
CC - was not sure if I was supposed to reply to the email or respond in the pull request itself for follow-up:

Thank you for taking the time to review my pull request, I have just pushed another commit to fix the issues you highlighted with one exception - with regard to Mobile Safari, I am unclear how to run the unit tests for a single particular browser. The instructions on https://github.com/prebid/Prebid.js/blob/master/CONTRIBUTING.md only go as far as to direct "gulp test", would you be able to clarify what command or arguments I should be using for the test you are requesting?

**Comment by mrcrawfo** [comment]
I have committed code to address the requested changes, so this is ready to be reviewed again. I do not seem to be able to edit the tags on this pull request but I believe its current status should be changed back to 'needs review'. Please let me know if there are any additional steps I need to take at this point to proceed with or resume the review process. Thank you!

**Comment by mrcrawfo** [comment]
@jaiminpanchal27 is there any update on the status of this pull request? I have submitted the requested changes but have not heard anything back in a while, can you please review those updates so we can continue moving this along towards approval/merging?

**Comment by jaiminpanchal27** [comment]
@mrcrawfo Thanks. Will check by EOD.

File: modules/lkqdBidAdapter.js [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+  if (serverResponse && serverResponse.body) {
+    if (serverResponse.error) {
+      utils.logError('Error: ' + serverResponse.error);
+      return bidResponses;
+    } else {
+      try {
+        let bidResponse = {};
+        if (bidRequest && bidRequest.data && bidRequest.data.bidId && bidRequest.data.bidId !== '') {
+          let sspXml;
+          try {
```
Comment (jaiminpanchal27) [inline modules/lkqdBidAdapter.js]: As per https://developer.mozilla.org/en-US/docs/Web/API/DOMParser on error handling.
Note that if the parsing process fails, the DOMParser does not throw an exception, but instead returns an error document:

```
<parsererror xmlns="http://www.mozilla.org/newlayout/xml/parsererror.xml">
(error description)
<sourcetext>(a snippet of the source XML)</sourcetext>
</parsererror>
```

Hence the try catch here does not make sense.
This code will still get executed and will result in error when parsing fails
```
bidResponse.creativeId = sspXml.getElementsByTagName('Ad')[0].getAttribute('id');
```

I think you can add something like https://stackoverflow.com/questions/11563554/how-do-i-detect-xml-parsing-errors-when-using-javascripts-domparser-in-a-cross to detect parsing error. This is just a suggestion, there maybe a better solution out there to detect dom parsing errors.

Also add a unit test where xml parsing fails.

Comment (mrcrawfo) [reply modules/lkqdBidAdapter.js]: Have just committed and pushed a fix to address this issue - thank you for catching this, as I was in fact unfamiliar with DOMParser's error handling and my implementation was expecting it to throw an error where it does not. Restructured my logic to account for this and added a new unit test for this case to demonstrate it working as requested.

Comment (jaiminpanchal27) [reply modules/lkqdBidAdapter.js]: @mrcrawfo I guess you forgot to push the code.

Comment (mrcrawfo) [reply modules/lkqdBidAdapter.js]: You are correct, I just pushed and it is working its way through CI now. Thank you for responding and calling this to my attention.

**Comment by jaiminpanchal27** [comment]
Please submit a PR to the [docs repo](https://github.com/prebid/prebid.github.io) to add a file for your adapter to the [bidders directory](https://github.com/prebid/prebid.github.io/tree/master/dev-docs/bidders) so your adapter's params will appear on the [bidders page](http://prebid.org/dev-docs/bidders.html). Thank you for contributing

**Comment by jaiminpanchal27** [comment]
https://github.com/prebid/prebid.github.io/pull/702

**Pull request merged**